File: 완주하지 못한 선수.cpp
#include <string>
#include <vector>
#include<iostream>
#include<algorithm>
using namespace std;

string solution(vector<string> participant, vector<string> completion) {
	 

	sort(participant.begin(), participant.end());
	sort(completion.begin(), completion.end());
	for (int i = 0; i < participant.size(); i++)
		cout << participant[i] <<" ";
	cout << "\n";
	for (int i = 0; i < completion.size(); i++)
		cout << completion[i] << " ";
	cout << "\n";

	for (int i = 0; i < completion.size(); i++) {
		if (participant[i] != completion[i]) {
			return participant[i];
		}
	}
	return participant.back();
}


int main() {
	vector<string> participant = { "mislav", "stanko", "mislav" ,"ana"};
	vector<string> completion = { "stanko", "ana","mislav" };
	cout<<solution(participant, completion);
}

File: 프로그래머스/level1/완주하지 못한 선수.js
function solution1 (participant, completion) {
    var answer = '';
    participant.sort()
    completion.sort()


    for (var i = 0; i < participant.length; i++) {
        if (participant[i] !== completion[i]) {
            answer = participant[i];
            return answer;
        }
    }
}

//hash로 푼 방법
function solution2 (participant, completion) {
    const hash = {};

    for (let val of participant) {
        if (!hash[val]) hash[val] = 0;
        hash[val]++;
    }
    console.log(hash)

    //완주한 선수는 제외
    completion.forEach(val => hash[val]--);

    //값이 남아있는 선수가 완주하지 못한 선수
    for (let key in hash) if (hash[key]) return key;
}
console.log(solution2(["mislav", "stanko", "mislav", "ana"], ["stanko", "ana", "mislav"]))